### ISSUE TITLE
JsonSerializer throws InvalidCastException

### ISSUE THREAD

**Issue by Fr1z2r** [dotnet/runtime#2387]
Using .NET Core App 3.1.101
This code throws an exception when IgnoreNullValues=true, and doesn't throw when IgnoreNullValues=false:
```
class MyClass
{
    public HashSet<int> X { get; set; } = new HashSet<int>();
}

public static void Main()
{
    JsonSerializer.Deserialize<MyClass>(@"{""X"":[2,1]}", 
        new JsonSerializerOptions {IgnoreNullValues = true});
}
```

```
System.InvalidCastException : Unable to cast object of type 'System.Collections.Generic.HashSet`1[System.Int32]' to type 'System.Collections.IList'.
   at System.Text.Json.JsonSerializer.ApplyObjectToEnumerable(Object value, ReadStack& state, Boolean setPropertyDirectly)
   at System.Text.Json.JsonSerializer.HandleEndArray(JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadCore(JsonSerializerOptions options, Utf8JsonReader& reader, ReadStack& readStack)
   at System.Text.Json.JsonSerializer.ReadCore(Type returnType, JsonSerializerOptions options, Utf8JsonReader& reader)
   at System.Text.Json.JsonSerializer.Deserialize(String json, Type returnType, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializer.Deserialize[TValue](String json, JsonSerializerOptions options)
```

**Comment by Jozkee** [comment]
This seems to be broken in 3.1 and 3.0. It was involuntary fixed for 5.0 under https://github.com/dotnet/corefx/pull/41482.

Repro steps: 
* Deserialize into properties that contain a default value, e.g. `MyClass` implemented above, using `IgnoreNullValues` option.
* Property needs to be of a type that does not implement `IList` but it has a constructor that can take an `IEnumerable` and that are cataloged as `ClassType.Enumerable` by the `JsonSerializer`.
* Affected types that I've seen so far:
  * `HashSet<T>`
  * `Stack<T>`
  * `Queue<T>`
  * `Stack`
  * `Queue`

cc @ahsonkhan @layomia

**Comment by layomia** [comment]
This is fixed in 5.0, and the workaround for 3.0/3.1 is to not initialize the collection. Does this work for you, or can you switch to 5.0?

Will close - this can be reopened if this is blocking on 3.1.
